ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoTransactionLogUpdatesSpec.scala

```diff
@@ -0,0 +1,174 @@
+import com.daml.lf.transaction.Node
+import com.daml.lf.transaction.Node.{KeyWithMaintainers, NodeCreate, NodeExercises}
+import com.daml.lf.value.Value
+import com.daml.platform.store.appendonlydao.events.{ContractId, NodeId}
+import com.daml.platform.store.entries.LedgerEntry
+import com.daml.platform.store.interfaces.TransactionLogUpdate
+import org.scalatest._
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.concurrent.Future
+
+private[dao] trait JdbcLedgerDaoTransactionLogUpdatesSpec
+    extends OptionValues
+    with Inside
+    with LoneElement {
+  this: AsyncFlatSpec with Matchers with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (getTransactionLogUpdates)"
+
+  it should "return only the ledger end marker if no new transactions in range" in {
+    for {
+      ledgerEnd @ (offset, eventSequentialId) <- ledgerDao.lookupLedgerEndOffsetAndSequentialId()
+      result <- transactionsOf(
+        ledgerDao.transactionsReader
+          .getTransactionLogUpdates(
+            startExclusive = ledgerEnd,
+            endInclusive = ledgerEnd,
+          )
+      )
+    } yield {
+      result should contain theSameElementsAs Seq(
+        TransactionLogUpdate.LedgerEndMarker(offset, eventSequentialId)
+      )
+      Succeeded
+    }
+  }
+
+  it should "return the correct transaction log updates" in {
+    for {
+      from <- ledgerDao.lookupLedgerEndOffsetAndSequentialId()
+      (offset1, t1) <- store(singleCreate)
+      (offset2, t2) <- store(txCreateContractWithKey(alice, "some-key"))
+      (offset3, t3) <- store(
+        singleExercise(
+          nonTransient(t2).loneElement,
+          Some(KeyWithMaintainers(someContractKey(alice, "some-key"), Set(alice))),
+        )
+      )
+      (offset4, t4) <- store(fullyTransient)
+      (offset5, t5) <- store(singleCreate)
+      (offset6, t6) <- store(singleNonConsumingExercise(nonTransient(t5).loneElement))
+      to <- ledgerDao.lookupLedgerEndOffsetAndSequentialId()
+      result <- transactionsOf(
+        ledgerDao.transactionsReader
+          .getTransactionLogUpdates(
+            startExclusive = from,
+            endInclusive = to,
+          )
+      )
+    } yield {
+      val (
+        Seq(
+          actualTx1: TransactionLogUpdate.Transaction,
+          actualTx2: TransactionLogUpdate.Transaction,
+          actualTx3: TransactionLogUpdate.Transaction,
+          actualTx4: TransactionLogUpdate.Transaction,
+          actualTx5: TransactionLogUpdate.Transaction,
+          actualTx6: TransactionLogUpdate.Transaction,
+        ),
+        Seq(endMarker: TransactionLogUpdate.LedgerEndMarker),
+      ) = result.splitAt(6)
+
+      val contractKey =
+        t2.transaction.nodes.head._2.asInstanceOf[NodeCreate[ContractId]].key.get.key
+      val exercisedContractKey = Map(offset2 -> contractKey, offset3 -> contractKey)
+
+      val eventSequentialIdGen = new AtomicLong(from._2 + 1L)
```

Comment:
Why do we need `AtomicLong` here? Is it for the handy get-and-increment method?

Reply:
Yes and most importantly: no other way in Scala to pass a mutable ref as a parameter otherwise